src/lua/LuaSerializer.cpp

```diff
@@ -462,17 +462,18 @@ void LuaSerializer::SaveComponents(Json &jsonObj, Space *space)
 
 	// Note: this loop relies on the ordering and contents of Space::m_bodies not changing
 	// between when bodies were serialized and when this function is called.
+	// Space should not do update, because it invalidates m_bodyIndex vector.
 
-	for (size_t idx = 0; idx < bodies.size(); idx++) {
-		Body *body = space->GetBodies()[idx];
+	for (size_t idx = 1; idx < bodies.size(); idx++) {
+		Body *body = space->GetBodyByIndex(idx);
```

Comment:
Rather than starting from a 1-based index, either update `GetBodyByIndex()` to operate on 0-based indices or call `GetBodyByIndex(idx + 1)` with a comment indicating why it takes 1-based indicies.

Reply:
This is done because `Space::RebuildBodyIndex()` and `Space::RebuildSystemBodyIndex()` is adding `nullptr` as first object at the start. When doing this PR I suggested that it's used as a "bad" index, but I don't know if it's really needed there.

Reply:
Changed every instance where start index was 1, instead of 0. I consider the first index of `Space::RebuildBodyIndex()` and `Space::RebuildSystemBodyIndex()` (aka `m_bodyIndex` and `m_sBodyIndex`) as a bad index.